### PR TITLE
Specify dtype to avoid Dataset size overflow on Windows

### LINKED ISSUE #901: Dataset size may overflow for large arrays on Windows
On Windows, `h5py.Dataset.size` may return negative values for large datasets:

```
>>> type(dataset)
h5py._hl.dataset.Dataset

>>> dataset.size
-1654967296

>>> dataset.shape
(7500000, 352)
```

This is caused by the default int being `int32` in numpy.
```
>>> import numpy
>>> numpy.prod((1000, 100, 1000, 100, 100))
-727379968
```
An easy fix would be to specify the `dtype` in `numpy.prod` when computing `size`.

### PR DESCRIPTION
**Pull request by kitchoi**
Fix #901

### CODE REVIEW

File: h5py/tests/hl/test_dataset_getitem.py [hunk 1]
```diff
@@ -452,18 +453,28 @@ def setUp(self):
...
         """ see issue #473 """
         self.assertNumpyBehavior(self.dset, self.data, np.s_[:,[0,1,2]])
 
-        
+
+class TestVeryLargeArray(TestCase):
+
+    def setUp(self):
+        TestCase.setUp(self)
+        self.data = np.empty((2**15, 2**16), dtype=np.bool)
```
Comment (tacaswell) [inline h5py/tests/hl/test_dataset_getitem.py]: You can create a data set without data, that is probably a better option 

```python

In [23]: d = F.create_dataset('huge3', (2**30, 2**30))

In [24]: d.size
Out[24]: 1152921504606846976

In [25]: d.shape
Out[25]: (1073741824, 1073741824)
```

**Comment by kitchoi** [comment]
Thanks @tacaswell 
I  just discovered there is a `test_dataset` under `h5py.tests.old`.  The test in this PR looks like it belongs there.  However the module is called `old`, so I am less sure because of that.  I will leave the test as it is now but I am happy to move it across if that's more appropriate.

The test that fails on CI (`h5py.tests.old.test_file.TestFileOpen`) seems to have nothing to do with `Dataset`, so I am not sure how it is related.

**Comment by aragilar** [comment]
That looks like we hit #888 (or at least something similar to it), I've restarted the appveyor test, let's see if it passes this time...

**Comment by tacaswell** [comment]
'power cycled' to re-run CI against current master.

**Pull request merged**